### PR TITLE
Support thread interrupting blocking functions (#1947)

### PR DESCRIPTION
**Pull request by jxdabc**
This is implementation of issue ~#57~ #1947 and non-intrusive variant of #1922

Signed-off-by: Trol <jiaoxiaodong@xiaomi.com>

### CODE REVIEW

**Comment by eggfly** [comment]
666

**Comment by jxdabc** [comment]
Amended as #1947:  
- Added an optional context parameter to enable single-call conversion of interruptible Java methods into main-safe suspending functions.
- Added some more tests.

**Comment by qwwdfsad** [comment]
https://github.com/Kotlin/kotlinx.coroutines/issues/1947#issuecomment-619365742
Here is our final design. 

Thanks a lot for your valuable contribution and moving this almost abandoned issue forward!
I realize there is a lot to do (along with a regular code-review that may take a bunch of iterations), so feel free to leave the rest of the changes on us if you don't have time to go this extra mile.

**Comment by jxdabc** [comment]
@qwwdfsad The design seems sweet. Especially as with avoid cross-block (launch...) misuse.

Thanks a lot for you to finally make coroutine together with interruption possible. That really helps a lot at least for some common Android patterns as I could see and make some people's work easy. I hope I could help complete this helpful progress.

File: kotlinx-coroutines-core/common/test/channels/ProduceTest.kt [hunk 1]
```diff
@@ -151,7 +151,7 @@ class ProduceTest : TestBase() {
             launch {
                 expect(2)
                 assertFailsWith<IllegalStateException> {
-                    awaitClose { expectUnreached() }
+                    awaitClose { expectUnreached() } // <- access outer ProducerScope from launch scope, won't compile
```
Comment (jxdabc) [inline kotlinx-coroutines-core/common/test/channels/ProduceTest.kt:154]: Existing code may need to use explicit reference to outer scope in this situation. 
`this@callbackFlow.awaitClose` in this case
Is that expected or is there something I missed?

**Comment by jxdabc** [comment]
@qwwdfsad 

I added DSL restriction just now with a draft commit (didn't amend the previous one). 

I found this DSL restriction may break existing code, as I commented as a code review comment for a example. Existing code may need change to use explicit reference to outer scope in this situation. Is that expected or is there something I missed?

File: kotlinx-coroutines-core/common/test/channels/ProduceTest.kt (hunk 1 above)
Comment (jxdabc) [reply kotlinx-coroutines-core/common/test/channels/ProduceTest.kt:154]: Just to mention that DslMarker will apply to all subclasses including ProducerScope (subclass of CoroutineScope)

File: kotlinx-coroutines-core/jvm/src/CancellationPoint.kt [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+                else -> throw IllegalStateException("unknown state")
+            }
+        }
+    }
+
+    fun clearInterrupt() {
+        state.loop { s ->
+            when {
+                s is Working -> if (state.compareAndSet(s, Finish)) { s.cancelHandle!!.dispose(); return }
+                s === Interrupting -> Thread.yield() // eases the thread
```
Comment (qwwdfsad) [inline kotlinx-coroutines-core/jvm/src/CancellationPoint.kt:114]: Please avoid `Thread.yield`. It is way too heavy-weight for 99% of usages.
Just spinning is enough

File: kotlinx-coroutines-core/jvm/src/CancellationPoint.kt [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+                val job = uCont.context[Job] ?: return@sc DefaultInterruptibleScope.block()
+                // slow path
+                val threadState = ThreadState().apply { initInterrupt(job) }
+                try {
+                    DefaultInterruptibleScope.block()
+                } finally {
+                    threadState.clearInterrupt()
+                }
+            } catch (e: InterruptedException) {
+                throw CancellationException()
```
[the remark this answers is not in the document]
Comment (jxdabc) [reply kotlinx-coroutines-core/jvm/src/CancellationPoint.kt]: There is no `CancellationException` constructor with a `cause` parameter on JVM, but still can be set with `initCause`.

File: kotlinx-coroutines-core/jvm/src/CancellationPoint.kt [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+private object DefaultInterruptibleScope : InterruptibleScope
+
+private class ThreadState {
+
+    fun initInterrupt(job: Job) {
+        // starts with Init
+        if (state.value !== Init) throw IllegalStateException("impossible state")
+        // remembers this running thread
+        state.value = Working(Thread.currentThread(), null)
```
[the remark this answers is not in the document]
Comment (jxdabc) [reply kotlinx-coroutines-core/jvm/src/CancellationPoint.kt]: We could get rid of `Init`. However, it seems we couldn't getting rid of `Init` AND starts with (Thread, cancelHandle). There is a race where `CancelHandler.invoke()` may arrive before starting. We could start with (Thread, null) instead.

File: kotlinx-coroutines-core/jvm/src/CancellationPoint.kt [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+                s is Working -> if (state.compareAndSet(s, Finish)) { s.cancelHandle!!.dispose(); return }
+                s === Interrupting -> Thread.yield() // eases the thread
+                s === Interrupted -> { Thread.interrupted(); return } // no interrupt leak
+                s === Init || s === Finish -> throw IllegalStateException("impossible state")
+                else -> throw IllegalStateException("unknown state")
+            }
+        }
+    }
+
+    private inner class CancelHandler : CompletionHandler {
```
[the remark this answers is not in the document]
Comment (jxdabc) [reply kotlinx-coroutines-core/jvm/src/CancellationPoint.kt]: Ya, it also seems meaningful to make ThreadState implement CompletionHandler.

**Comment by jxdabc** [comment]
@qwwdfsad 
Thanks a lot for detailed advice. I amended the patch as code review comments.

**Comment by jxdabc** [comment]
Thanks! Glad to meet you all.

**Comment by qwwdfsad** [comment]
Merged via #1972.
Thanks!

**Pull request closed without merging**